### PR TITLE
Document std::os::raw.

### PR DESCRIPTION
**Pull request by clarfonthey**
This adds a brief explanation to each type and its definition according to C. This also helps clarify that the definitions of the types, as described by rustdoc, are not necessarily the same from platform to platform.

### CODE REVIEW

File: src/libstd/os/raw/mod.rs [hunk 1]
```diff
@@ -36,30 +44,46 @@ use fmt;
...
+/// Equivalent to C's `void` type when used as a [pointer].
 ///
-/// This type is only useful as a pointer target. Do not use it as a
-/// return type for FFI functions which have the `void` return type in
-/// C. Use the unit type `()` or omit the return type instead.
+/// In essence, `*const c_void` is equivalent to C's `const void*`
+/// and `*mut c_void` is equivalent to C's `void*`. That said, this is
+/// *not* the same as C's `void` return type, which is Rust's `()` type.
+///
+/// [pointer]: ../primitive.pointer.html
```
Comment (kennytm) [inline src/libstd/os/raw/mod.rs]: Broken link. Needs two `../` I guess.

```
[01:21:17] std/os/raw/index.html:66: broken link - std/os/primitive.pointer.html
[01:21:17] std/os/raw/enum.c_void.html:56: broken link - std/os/primitive.pointer.html
[01:21:26] thread 'main' panicked at 'found some broken links', /checkout/src/tools/linkchecker/main.rs:49:8
```

File: src/libstd/os/raw/char.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Equivalent to C's `char` type.
+
+[C's `char` type] is completely unlike [Rust's `char` type]; while Rust's type represents a unicode scalar value, C's `char` type is just an ordinary integer. In practice, this type will always be either [`i8`] or [`u8`], but you're technically not supposed to rely on this behaviour, as the standard only defines a char as being at least eight bits long.
+
+C chars are most commonly used to make C strings. Unlike Rust, where the length of a string is included alongside the string, C strings mark the end of a string with a zero. See [`CStr`] for more information.
```
Comment (retep998) [inline src/libstd/os/raw/char.md]: >C strings mark the end of a string with a zero

This is slightly misleading as someone might think it means `0` the character, not the literal integer value 0.

File: src/libstd/os/raw/long.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Equivalent to C's `signed long` (`long`) type.
+
+This type will usually be [`i64`], but is sometimes [`i32`] \(i.e. [`isize`]\) on 32-bit systems. Technically, the standard only requires that it be at least 32 bits, or at least the size of an [`int`].
```
Comment (retep998) [inline src/libstd/os/raw/long.md]: Also note that it is equivalent to `i32` even on 64bit Windows.

Comment (clarcharr) [reply src/libstd/os/raw/long.md]: I was not aware of this; I'll remove the remarks about `isize` and `usize` in these docs.

File: src/libstd/os/raw/mod.rs (hunk 1 above)
Comment (clarcharr) [reply src/libstd/os/raw/mod.rs]: I double-checked every link but this one. ><

File: src/libstd/os/raw/char.md (hunk 2 above)
Comment (clarcharr) [reply src/libstd/os/raw/char.md]: Would clarifying "the character '\0' " be sufficient?

**Comment by clarcharr** [comment]
Made revisions per comments

File: src/libstd/os/raw/schar.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+Equivalent to C's `signed char` type.
+
+This type will almost always be [`i8`], but its size is technically equal to the size of a C [`char`], which isn't very clear-cut.
```
Comment (QuietMisdreavus) [inline src/libstd/os/raw/schar.md]: The use of "isn't very clear-cut" throughout the number docs feels awkward. Is there *any* guidance we can paraphrase here? If not, i'd rather we use a phrase like "which is left under-specified" to make it feel marginally less wishy-washy.

Comment (clarcharr) [reply src/libstd/os/raw/schar.md]: Fair; I was trying to avoid duplicating the docs too much, although it's mostly wishy-washy because C's definitions are incredibly wishy-washy.

I'll probably replace it with "which does not have a strict definition like Rust does" or something a bit more eloquent.

Comment (kennytm) [reply src/libstd/os/raw/schar.md]: I don't see how C doesn't have a strict definition of size of `signed char` 😄. In C a `char`'s size is defined to be 1 byte, always (although 1 byte is not defined to be 8 bits) (C11 §6.5.3.4/4). 

Unless Rust is going to support non-8-bit-byte system, a `signed char` is always an `i8`.

**Comment by kennytm** [comment]
Hi @clarcharr, just checking if this PR is still actively worked on. Could you rephrase the sentence in https://github.com/rust-lang/rust/pull/46962#pullrequestreview-86965062? Thanks!

**Comment by clarcharr** [comment]
@kennytm will do in a bit!

**Comment by shepmaster** [comment]
Ping from triage, @clarcharr ! We haven't heard from you in a week; will you still be able to make the most recent set of changes?

**Comment by clarcharr** [comment]
@shepmaster + @kennytm + @QuietMisdreavus my laptop was out of commission for a bit and I'm working on rewording things right now.

**Comment by clarcharr** [comment]
I've pushed the latest version; let me know if things are okay!

File: src/libstd/os/raw/uint.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Equivalent to C's `unsigned int` type.
+
+This type will almost always be [`u16`], but may differ on some esoteric systems. The C standard technically only requires that this type be an unsigned integer with the same size as an [`int`]; some systems define it as a [`u16`], for example.
```
Comment (QuietMisdreavus) [inline src/libstd/os/raw/uint.md]: Is that first `u16` supposed to be a `u32`?

Comment (clarcharr) [reply src/libstd/os/raw/uint.md]: ...yes

**Comment by pietroalbini** [comment]
Ping from triage @QuietMisdreavus!

**Comment by QuietMisdreavus** [comment]
I was still waiting on one last change, but it was small enough that i put it in myself. (Sorry @clarcharr! `>_>`)

@bors r+

**Comment by clarcharr** [comment]
@quietmisdreavus I actually think I have the commit on my machine but I forgot to push ><
